repo2docker/buildpacks/r.py

```diff
@@ -64,21 +71,22 @@ def detect(self):
         unless a `requirements.txt` is present and we do not want to require the
         presence of a `requirements.txt` to use R.
 
-        Instead we just check if runtime.txt contains a string of the form
-        `r-<YYYY>-<MM>-<DD>`
+        Instead we check the options described in the class comment above.
         """
         # If no date is found, then self.checkpoint_date will be False
         # Otherwise, it'll be a date object, which will evaluate to True
         if self.checkpoint_date:
             return True
+
         description_R = 'DESCRIPTION'
-        if not os.path.exists('binder') and os.path.exists(description_R):
-            if not self.checkpoint_date:
-                # no R snapshot date set through runtime.txt
-                # set the R runtime to the latest date that is guaranteed to be on MRAN across timezones
-                self._checkpoint_date = datetime.date.today() - datetime.timedelta(days=2)
-                self._runtime = "r-{}".format(str(self._checkpoint_date))
-            return True
+        if not os.path.exists('binder'): 
+            if os.path.exists(description_R) or (self.stencila_manifest_contexts and "r" in self.stencila_manifest_contexts):
```

Comment:
The reason that `DESCRIPTION` files are only recognised if there is no `binder/` sub-directory is that we want to allow "opportunistic" building of R packages from their Git repository but allow package maintainers to override that by creating a `binder/` directory which specifies how to install things.

Is this what users of stencila expect? I think of stencila as a different frontend (jupyter, RStudio, nteract, stencila) or a viewer for certain types of files. So I think we still need to let users specify dependencies that should be installed which they might want to keep tidy by putting them in a `binder/` directory. What do you think?

Reply:
I thought it's better to align the Stencila workflow with other use cases, i.e. if a `/binder` directory is there, no "repo2docker" magic happens. But I must admit I have have misunderstood the use of the `binder/` directory! So you are right, this change is not useful and Stencila users should be able to put their configurations into a `binder/` directory.

As for _Stencila users_, the whole project is aimed at non-coders, so I think it is unlikely that we have an author of a Stencila document who wants to write her own Dockerfile. An install.R script is probably as far as they may want to go.

_Should this also mean we should _not_ stop detecting `manifest.xml` and the executions contexts if there is a `binder/` directory?_

I think so! Because if we stop the detection, a user must install all the Stencila stuff herself, which means that particular user may also define a `Dockerfile`, see new example at https://github.com/nuest/stencila-multi (still checking if that works as advertised).

I think this is more in line with the "regular way" of using binder, and not mimicking the R package use case. Thanks for pointing this out @betatim

Reply:
If there is a `binder/` subdirectory repo2docker will still spring into action. It will however only consider files in this subdirectory.

If there is a `Dockerfile` then repo2docker follows that and does nothing beyond `docker build`.

I think we should ignore the presence or absence of a `binder/` directory for deciding things for stencila.

Detect which languages are needed by looking at all the manifest files and then use the composability of repo2docker to let people install extra libraries via `requirements.txt` or `install.R` if they want to.

Reply:
Should be implemented this way now.